geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/QueryStrategyDecider.scala

```diff
@@ -78,51 +67,35 @@ object QueryStrategyDecider {
    */
   private def processAnd(and: And, sft: SimpleFeatureType, hints: StrategyHints): Strategy = {
 
-    val filters: util.List[Filter] = decomposeAnd(and)
+    import org.locationtech.geomesa.utils.geotools.RichIterator.RichIterator
+
+    val filters = decomposeAnd(and)
 
     // scan the query and identify the type of predicates present
 
     // record strategy takes priority
-    val recordStrategies =
-      filters.toStream.flatMap(f => getRecordIdxStrategy(f, sft).map(StrategyAndFilter(_, f)))
-    if (!recordStrategies.isEmpty) {
-      return recordStrategies(0).strategy
+    val recordStrategy = filters.iterator.flatMap(f => RecordIdxStrategy.getStrategy(f, sft, hints)).headOption
+    if (recordStrategy.isDefined) {
+      return recordStrategy.get.strategy
     }
 
+    // look for reasonable cost attribute strategies
     val attributeStrategies =
-      filters.flatMap(f => getAttributeIndexStrategy(f, sft).map(StrategyAndFilter(_, f)))
+      filters.flatMap(f => AttributeIndexStrategy.getStrategy(f, sft, hints)).filter(_.cost < REASONABLE_COST)
     // if no attribute or record strategies, use ST
     if (attributeStrategies.isEmpty) {
       return new STIdxStrategy
     }
 
-    // next look for high-cardinality attribute filters
-    val highCardinalityStrategy = attributeStrategies.find { case StrategyAndFilter(strategy, filter) =>
-      val (prop, _) = AttributeIndexStrategy.getPropertyAndRange(filter, sft)
-      hints.cardinality(sft.getDescriptor(prop)) == Cardinality.HIGH
-    }
+    // next look for low cost (high-cardinality) attribute filters - cost is set in the attribute strategy
+    val highCardinalityStrategy = attributeStrategies.find(_.cost < OPTIMAL_COST)
     if (highCardinalityStrategy.isDefined) {
       return highCardinalityStrategy.get.strategy
     }
 
-    // finally, compare spatial and attribute filters based on order
-    val stStrategy = filters.flatMap(f => getSTIdxStrategy(f, sft).map(StrategyAndFilter(_, f))).headOption
-    val attrStrategy = attributeStrategies.find { case StrategyAndFilter(strategy, filter) =>
-      val (prop, _) = AttributeIndexStrategy.getPropertyAndRange(filter, sft)
-      hints.cardinality(sft.getDescriptor(prop)) != Cardinality.LOW
-    }
-
-    (stStrategy, attrStrategy) match {
-      case (None, None)           => new STIdxStrategy
-      case (Some(st), None)       => st.strategy
-      case (None, Some(attr))     => attr.strategy
-      case (Some(st), Some(attr)) =>
-        if (filters.indexOf(st.filter) < filters.indexOf(attr.filter)) {
-          st.strategy
-        } else {
-          attr.strategy
-        }
-    }
+    // finally, prefer spatial filters if available
+    val stStrategy = filters.iterator.flatMap(f => STIdxStrategy.getStrategy(f, sft, hints)).headOption
+    stStrategy.orElse(attributeStrategies.headOption).map(_.strategy).getOrElse(new STIdxStrategy)
```

Comment:
I refactored this a bit per anthony's comments. Let me know once I push the changes if you think it could still use some clarification.